crypto/eth2_hashing/src/lib.rs

```diff
@@ -1,61 +1,210 @@
-//! Provides a simple hash function utilizing `ring::digest::SHA256`.
+//! Optimized SHA256 for use in Ethereum 2.0.
 //!
-//! The purpose of this crate is to provide an abstraction to whatever hash function Ethereum
-//! 2.0 is using. The hash function has been subject to change during the specification process, so
-//! defining it once in this crate makes it easy to replace.
-
-#[cfg(not(target_arch = "wasm32"))]
-pub use ring::digest::{digest, Context, Digest, SHA256};
+//! The initial purpose of this crate was to provide an abstraction over the hash function used in
+//! Ethereum 2.0. The hash function changed during the specification process, so defining it once in
+//! this crate made it easy to replace.
+//!
+//! Now this crate serves primarily as a wrapper over two SHA256 crates: `sha2` and `ring` –
+//! which it switches between at runtime based on the availability of SHA intrinsics.
 
-#[cfg(target_arch = "wasm32")]
-use sha2::{Digest, Sha256};
+pub use self::DynamicContext as Context;
+use sha2::Digest;
 
 #[cfg(feature = "zero_hash_cache")]
 use lazy_static::lazy_static;
 
-/// Returns the digest of `input`.
-///
-/// Uses `ring::digest::SHA256`.
-pub fn hash(input: &[u8]) -> Vec<u8> {
-    #[cfg(not(target_arch = "wasm32"))]
-    let h = digest(&SHA256, input).as_ref().into();
-
-    #[cfg(target_arch = "wasm32")]
-    let h = Sha256::digest(input).as_ref().into();
+/// Length of a SHA256 hash in bytes.
+pub const HASH_LEN: usize = 32;
 
-    h
+/// Returns the digest of `input` using the best available implementation.
+pub fn hash(input: &[u8]) -> Vec<u8> {
+    DynamicImpl::best().hash(input)
 }
 
-/// Compute the hash of two slices concatenated.
-///
-/// # Panics
+/// Hash function returning a fixed-size array (to save on allocations).
 ///
-/// Will panic if either `h1` or `h2` are not 32 bytes in length.
-#[cfg(not(target_arch = "wasm32"))]
-pub fn hash32_concat(h1: &[u8], h2: &[u8]) -> [u8; 32] {
```

Comment:
I notice that `has32_concat` on not-WASM will allocate a `[u8; 64]` where it didn't previously. It *seems* like it would have a very small effect, but I'm wondering if you looked into this.

Reply:
Oh yeah, good catch. We may as well skip that and feed the bytes directly into a context. I'll update shortly